cmd/gitserver/server/common/queue.go

```diff
@@ -3,45 +3,154 @@ package common
+	//
+	// If it is invoked before SetPushedAt (see below), it will return the zero value of time.Time.
+	GetPushedAt() time.Time
+
+	// SetPushedAt sets the pushedAt time on the JobMetadata. It is idempotent.
+	//
+	// Invoking this the first time will set the current time and subsequent calls to this method
+	// will be a NOOP.
+	SetPushedAt()
+}
+
+// JobMetadata provides an API for managing metadata of a job. This is useful in metrics collection.
+//
+// Implementations of Jobber should embed this struct.
+type JobMetadata struct {
+	pushedAt     time.Time
+	pushedAtOnce sync.Once
+}
+
+func (jm *JobMetadata) GetPushedAt() time.Time {
+	return jm.pushedAt
+}
+func (jm *JobMetadata) SetPushedAt() {
+	jm.pushedAtOnce.Do(func() {
+		jm.pushedAt = time.Now()
+	})
+}
+
 // Queue is a threadsafe FIFO queue.
-type Queue[T any] struct {
-	mu   sync.Mutex
+type Queue[T Jobber] struct {
+	*metrics
+
 	jobs *list.List
 
+	mu sync.Mutex
+
+	// FIXME: Make these private.
+	// Coming soon in a follow up PR.
 	Mutex sync.Mutex
 	Cond  *sync.Cond
 }
 
 // NewQueue initializes a new Queue.
-func NewQueue[T any](jobs *list.List) *Queue[T] {
+func NewQueue[T Jobber](obctx *observation.Context, name string, jobs *list.List) *Queue[T] {
 	q := Queue[T]{jobs: jobs}
 	q.Cond = sync.NewCond(&q.Mutex)
 
+	// Register the metrics the first time this queue is used.
+	registerMetrics(obctx)
+
+	// Setup the metrics for this specific instance of the queue.
+	q.metrics = &metrics{
+		length:         queueLength.WithLabelValues(name),
+		enqueuedTotal:  queueEnqueuedTotal.WithLabelValues(name),
+		waitTime:       queueWaitTime.WithLabelValues(name),
+		processingTime: queueProcessingTime.WithLabelValues(name),
+	}
+
 	return &q
 }
 
 // Push will queue the job to the end of the queue.
-func (q *Queue[T]) Push(job *T) {
+func (q *Queue[T]) Push(job T) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
 	q.jobs.PushBack(job)
 	q.Cond.Signal()
+
+	// Set the push time on the job's metadata. This will be used to observe the total wait time in
+	// queue when this job is eventually popped.
+	job.SetPushedAt()
+	q.length.Inc()
+	q.enqueuedTotal.Inc()
 }
 
-// Pop will return the next job. If there's no next job available, it returns nil.
-func (q *Queue[T]) Pop() *T {
+// Pop returns the next job and a function that consumers of this job may use to record some
+// metrics. If there's no next job available, it returns nil, nil.
+func (q *Queue[T]) Pop() (*T, func(float64)) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
 	next := q.jobs.Front()
 	if next == nil {
-		return nil
+		return nil, nil
 	}
 
-	return q.jobs.Remove(next).(*T)
+	job := q.jobs.Remove(next).(T)
+
+	q.waitTime.Observe(time.Since(job.GetPushedAt()).Seconds())
+	q.length.Dec()
+
+	// NOTE: The function being returned is hardcoded at the moment. In the future this may be a
+	// property of the queue if implementations need it. For now this is all we need.
+	return &job, func(val float64) {
+		q.processingTime.Observe(val)
+	}
```

Comment:
if I am not mistaken you don't need to pass in a val. You could do this
```suggestion
	processingStart := time.Now()
	return &job, func() {
		q.processingTime.Observe(time.Since(processingStart).Seconds())
	}
```

Reply:
Yeah, that makes sense. I just tested it out with this:

```go
package main

import (
	"fmt"
	"time"
)

func main() {
	start := time.Now()
	slowFunc := func() {
		fmt.Println("starting slow func")
		time.Sleep(2 * time.Second)
		fmt.Println(time.Since(start).Seconds())
	}

	fmt.Println("since start", time.Since(start).Seconds())
	slowFunc()
}
```

And the output is:

```
$ go run main.go
since start 4.2e-08
starting slow func
2.000240834
```

Reply:
I've fixed this now. But the function now returns the duration elapsed so that consumers know how long it took and can decide to do something based on that information if they want.